lib/motsu/src/context.rs

```diff
@@ -1,34 +1,98 @@
 //! Unit-testing context for Stylus contracts.
-use std::sync::{Mutex, MutexGuard};
 
+use std::{collections::HashMap, ptr};
+
+use dashmap::DashMap;
+use once_cell::sync::Lazy;
 use stylus_sdk::{alloy_primitives::uint, prelude::StorageType};
 
-use crate::storage::reset_storage;
+use crate::prelude::{Bytes32, WORD_BYTES};
 
-/// A global static mutex.
-///
-/// We use this for scenarios where concurrent mutation of storage is wanted.
-/// For example, when a test harness is running, this ensures each test
-/// accesses storage in an non-overlapping manner.
+/// Context of stylus unit tests associated with the current test thread.
+#[allow(clippy::module_name_repetitions)]
+pub struct TestContext {
+    thread_name: TestThreadName,
+}
+
+impl TestContext {
+    /// Get test context associated with the current test thread.
+    #[must_use]
+    pub fn current() -> Self {
+        Self { thread_name: TestThreadName::current() }
+    }
+
+    /// Get the value at `key` in storage.
+    pub(crate) fn get_bytes(self, key: &Bytes32) -> Bytes32 {
+        let storage = MOCK_STORAGE.entry(self.thread_name).or_default();
+        storage.contract_data.get(key).copied().unwrap_or_default()
+    }
+
+    /// Get the raw value at raw `key` in storage.
+    pub(crate) unsafe fn get_bytes_raw(self, key: *const u8, value: *mut u8) {
+        let key = read_bytes32(key);
+
+        write_bytes32(value, self.get_bytes(&key));
+    }
+
+    /// Set the value at `key` in storage to `val`.
+    pub(crate) fn set_bytes(self, key: Bytes32, val: Bytes32) {
+        let mut storage = MOCK_STORAGE.entry(self.thread_name).or_default();
+        storage.contract_data.insert(key, val);
+    }
+
+    /// Set the raw value at `key` in storage to raw `val`.
+    pub(crate) unsafe fn set_bytes_raw(self, key: *const u8, value: *const u8) {
+        let (key, value) = (read_bytes32(key), read_bytes32(value));
+        self.set_bytes(key, value);
+    }
+
+    /// Clears storage, removing all key-value pairs associated with the current
+    /// test thread.
+    pub fn reset_storage(self) {
+        MOCK_STORAGE.remove(&self.thread_name);
+    }
+}
+
+/// Storage mock: A global mutable key-value store.
+/// Allows concurrent access.
 ///
-/// See [`with_context`].
-pub(crate) static STORAGE_MUTEX: Mutex<()> = Mutex::new(());
-
-/// Acquires access to storage.
-pub fn acquire_storage() -> MutexGuard<'static, ()> {
-    STORAGE_MUTEX.lock().unwrap_or_else(|e| {
-        reset_storage();
-        e.into_inner()
-    })
+/// The key is the name of the test thread, and the value is the storage of the
+/// test case.
+static MOCK_STORAGE: Lazy<DashMap<TestThreadName, MockStorage>> =
```

Comment:
The same as https://github.com/OpenZeppelin/rust-contracts-stylus/pull/379/files#r1815499126 in terms of `STORAGE`

Reply:
Yeah, static var `MOCK_STORAGE` colliding with `MockStorage` struct. `STORAGE` name will make sense